### PR TITLE
Async support on the redesign branch

### PR DESCRIPTION
**Pull request by mariuszjamro**
Async is the state of the art way of doing things and previous (current?) version of FluentScheduler doesn't handle async natively (and it seems it won't #117). I propose following changes to the redesign branch to address that:

1. Schedule should accept Actions returning Task. This allows native async jobs to be run/created easily.

```
var schedule = new Schedule(
    async () => await Console.Out.WriteLineAsync("5 minutes just passed."),
    run => run.Every(5).Minutes()
);
```

2. `Schedule` should accept Actions with CancellationToken as an argument. This allows jobs to be notified when scheduler is stopped and end work sooner if possible.

```
var schedule = new Schedule(
    async (CancellationToken cancellationToken) =>
    {

        for(var i = 0; i < 100; i++)
        {
            await Console.Out.WriteLineAsync("Im doing something");
            await Task.Delay(1000);
            if (cancellationToken.IsCancellationRequested)
            {
                // Schedule cancelled
                return;
            }
        }
    },
    run => run.Every(5).Minutes()
);
```

### CODE REVIEW

File: FluentScheduler.UnitTests/Scheduler/ScheduleTests.cs [hunk 1]
```diff
@@ -262,5 +306,25 @@ public void UseUtcAfterStop()
...
+
+        [TestMethod]
+        public void WaitForCancellation()
+        {
+            // Arrange
+            var cancelled = false;
+            var schedule = new Schedule(async (cancellationToken) =>
+            {
+                cancellationToken.WaitHandle.WaitOne(1000);
+                cancelled = cancellationToken.IsCancellationRequested;
```
Comment (tallesl) [inline FluentScheduler.UnitTests/Scheduler/ScheduleTests.cs:318]: Minor thing: I believe `cancelled = cancellationToken.WaitHandle.WaitOne(1000)` would do the same.

File: FluentScheduler/Scheduler/Schedule.cs [hunk 2]
```diff
@@ -17,17 +17,68 @@ public class Schedule
         /// </summary>
         /// <param name="job">Job to be scheduled</param>
         /// <param name="cronExpression">The scheduling as a cron expression</param>
-        public Schedule(Action job, string cronExpression) =>
+        public Schedule(Action job, string cronExpression) : this(() => MakeAsync(job), cronExpression)
+        {
+        }
```
Comment (tallesl) [inline FluentScheduler/Scheduler/Schedule.cs:22]: Minor thing: rest of the project puts empty blocks on the same line (`Foo() { }`).

File: FluentScheduler/Scheduler/Schedule.cs [hunk 3]
```diff
@@ -17,17 +17,68 @@ public class Schedule
...
+        public Schedule(Action job, Action<RunSpecifier> specifier) : this(() => MakeAsync(job), specifier)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new schedule for the given job.
+        /// </summary>
+        /// <param name="job">Job to be scheduled</param>
+        /// <param name="cronExpression">The scheduling as a cron expression</param>
+        public Schedule(Func<Task> job, string cronExpression) : this((ct) => job(), cronExpression)
```
Comment (tallesl) [inline FluentScheduler/Scheduler/Schedule.cs:38]: Minor thing: rest of the project calls it `token`.

File: FluentScheduler/Scheduler/Schedule.cs [hunk 4]
```diff
@@ -17,17 +17,68 @@ public class Schedule
...
+        /// <param name="action"></param>
+        /// <returns></returns>
+        private static Task MakeAsync(Action action)
+        {
+            action();
+
+            // On .NET Standard 2 it should return the Task.CompletedTask singleton to avoid creating new task each time.
+            // Because the library is using .NET Standard 1 we mimic completed task with a Task returning any result
+            // (the result is not used anway).
+            return Task.FromResult(0);
```
Comment (tallesl) [inline FluentScheduler/Scheduler/Schedule.cs:79]: We're using net standard 2 now. Should be ok to change.

File: FluentScheduler/Scheduler/InternalSchedule.cs [hunk 5]
```diff
@@ -78,16 +78,20 @@ internal void Stop(bool block, int? timeout = null)
...
-            _tokenSource = null;
+                if (block && !timeout.HasValue)
+                    _task.Wait();
+            }
+            finally
+            {
+                _tokenSource.Dispose();
+                _task = null;
+                _tokenSource = null;
+            }
```
Comment (tallesl) [inline FluentScheduler/Scheduler/InternalSchedule.cs:94]: Good catch. I might apply this change in a separate commit and test it alone before merging this pr.

**Comment by tallesl** [comment]
I'm just getting back to work on the library, sorry for the long delay.

I've been pushing against the async wrap in the past, but now it's the right time to fix that. Thank you for the pull request, from a first glance looks perfect.

I added some comments, but they are more of reminders to myself when merging it, you don't have to change a thing.

File: FluentScheduler/Scheduler/Schedule.cs (hunk 3 above)
Comment (mariuszjamro) [reply FluentScheduler/Scheduler/Schedule.cs]: Fixed.

File: FluentScheduler.UnitTests/Scheduler/ScheduleTests.cs (hunk 1 above)
Comment (mariuszjamro) [reply FluentScheduler.UnitTests/Scheduler/ScheduleTests.cs]: Fixed.

**Comment by tallesl** [comment]
Merged, available on 6.0.0-alpha.3.

Since I performed a rebase, I kindly ask you to throw away your current fork (and create a fresh one if you need to).

Thank you for the contribution.

**Pull request closed without merging**